content/blog/multi-single-tenant-github-app-installation/index.md

```diff
@@ -0,0 +1,62 @@
+
+![multi-single-tenant.png](multi-single-tenant.png)
+
+# Requirements
+
+Roadie provides a single GitHub App on the GitHub App marketplace called “Roadie Backstage”.
+
+The requirements we had around GitHub Apps was as follows:
+- The admin user on-boarding onto a newly created roadie backstage instance must be hinted to install the “Roadie Backstage” App
+- There must be absolute isolation between installations of the App. E.g., customer 1 must not be able to access the resources of customer 2 via Roadie.
+
+# Routing the Callback URL Redirect
+
+GitHub Apps provide a single “Callback URL” configuration option. When this is set, the user installing the backstage app is redirected to the configured Callback URL.
+
+Because there is one Callback URL we had to create a single service shared by all customers that could be used to handle the installation redirect. So we did just that. We created a lambda API service that we called the “redirect-handler”. It would accept a “state” parameter that contains the URL target of the redirect. The redirect handler would simply unpack the state parameter and return the user a redirect to the target, reproducing all of the available url params contained in the request.
+
+![redirect-handler.png](redirect-handler.png)
+
+As you might be able to decern by, the state variable is what the redirect-handler uses to decide where to send the user on the redirect. This state is sent to the user by the backstage application, when the user indicates that they want to install the GitHub App. The state variable is a JWT token signed by a private key owned by the Backstage instance. The redirect-handler is configured with the required trust store to trust JWT tokens signed by the customer tenants. The JWT tokens contain a payload item called "urn:roadie:rdt".
+
+![app-installation-flow.png](app-installation-flow.png)
+
+When the redirect-handler recieves the request, it is decoding the JWT token in the state variable, verifies it is signed by a trusted source, extracts the "urn:roadie:rdt" variable and sends the user back a redirect to there.
+
+Now that the customers backstage instance has received the redirect, Backstage must verify that it trusts this installation.
+
+# Trusting the Installation ID returned via a redirect
+
+The customer backstage instance receives a request like the following, after the redirect hops have completed.
+
+```javascript
+GET - https://customern.roadie.so/github-installation?state=<jwt>&code=1234&installation_id=123456&setup_action=install
+```
+
+Roadie cannot just trust that the customer has provided an installation id that the user has ownership of. As such, Backstage must verify that the user is allowed to install this app. We have done this by enabling the feature "Request user authorization (OAuth) during installation" in GitHub Apps
```

Comment:
There's a mention of  both roadie/backstage here. Do you think we should just choose one?

I know in the past when I was writing docs for the k8s stuff David had suggested to use Roadie rather than Backstage.

Reply:
Could you take a look again. When I say backstage, I am specifically meaning the backstage application. When I say roadie, I mean the company.